onnxruntime/core/providers/webnn/builders/impl/rotaryEmbedding_op_builder.cc

```diff
@@ -0,0 +1,314 @@
+  emscripten::val concated_partial_input0 = wnn_builder.call<emscripten::val>(
+      "concat", split_partial_input0.call<emscripten::val>("reverse"), split_axis, concat_partial_input0_options);
+
+  if (position_ids_is_offset) {
+    // We generate a sequence from 0 to sequence_length and add the offset to it.
+    const std::vector<uint32_t> position_ids_range_shape = {1, sequence_length};
+    emscripten::val position_ids_range_buffer = emscripten::val::global("BigInt64Array").new_(sequence_length);
+    for (uint32_t i = 0; i < sequence_length; i++) {
+      position_ids_range_buffer.set(i, emscripten::val::global("BigInt")(i));
+    }
+    emscripten::val position_ids_range_desc = emscripten::val::object();
+    position_ids_range_desc.set("shape", emscripten::val::array(position_ids_range_shape));
+    position_ids_range_desc.set("dimensions", emscripten::val::array(position_ids_range_shape));
+    position_ids_range_desc.set("dataType", emscripten::val("int64"));
+    emscripten::val position_ids_range = wnn_builder.call<emscripten::val>(
+        "constant", position_ids_range_desc, position_ids_range_buffer);
+    // Add the offset to the sequence.
+    emscripten::val position_ids_add_range_options = emscripten::val::object();
+    position_ids_add_range_options.set("label", node_name + "_position_ids_add_range");
+    position_ids = wnn_builder.call<emscripten::val>(
+        "add", position_ids, position_ids_range, position_ids_add_range_options);
+  }
+
+  // Gather the cos/sin values based on the position_ids.
+  emscripten::val gather_cos_sin_options = emscripten::val::object();
+  gather_cos_sin_options.set("label", node_name + "_gather_cos_sin");
+  gather_cos_sin_options.set("axis", 0);
+  emscripten::val gather_cos = wnn_builder.call<emscripten::val>(
+      "gather", cos_cache, position_ids, gather_cos_sin_options);
+  emscripten::val gather_sin = wnn_builder.call<emscripten::val>(
+      "gather", sin_cache, position_ids, gather_cos_sin_options);
+
+  // After gathering cos/sin, reshape and broadcast them to match the number of heads of the input data.
+  const std::vector<uint32_t> reshaped_cos_sin_shape =
+      interleaved ? std::vector<uint32_t>({batch_size, sequence_length, 1, rotary_embedding_dim / 2, 1})
+                  : std::vector<uint32_t>({batch_size, sequence_length, 1, 1, rotary_embedding_dim / 2});
+  emscripten::val reshape_gather_cos_sin_options = emscripten::val::object();
+  reshape_gather_cos_sin_options.set("label", node_name + "_reshape_gather_cos_sin");
+  gather_cos = wnn_builder.call<emscripten::val>(
+      "reshape", gather_cos, emscripten::val::array(reshaped_cos_sin_shape), reshape_gather_cos_sin_options);
+  gather_sin = wnn_builder.call<emscripten::val>(
+      "reshape", gather_sin, emscripten::val::array(reshaped_cos_sin_shape), reshape_gather_cos_sin_options);
+
+  // Multiply the non-roated data with the cos and the rotated data with the sin.
```

Comment:
non-rotated

Also suggest spelling "cosine" and "sine" when in full English writing like comments (but okay not to too).

Reply:
Thanks, done.